AT/AT/docs/man/man3/DRV_GENERAL.3
.TH "DRV_GENERAL" 3 "Mon Nov 12 2018" "Lab TD" \" -*- nroff -*-
.ad l
.nh
.SH NAME
DRV_GENERAL \- 
.PP
General structures, types, functions and defines\&.  

.SS "Data Structures"

.in +1c
.ti -1c
.RI "struct \fBUART_BAUDRATE_VALUE_DLD\fP"
.br
.ti -1c
.RI "struct \fBUART_BAUDRATE_VALUE_DIV_MUL\fP"
.br
.in -1c
.SS "Macros"

.in +1c
.ti -1c
.RI "#define \fBMHZ\fP   *1000000l"
.br
.ti -1c
.RI "#define \fBKHZ\fP   *1000l"
.br
.ti -1c
.RI "#define \fBHZ\fP   *1l"
.br
.ti -1c
.RI "#define \fBMAX\fP(x, y)   ( (x > y) ? x : y )"
.br
.ti -1c
.RI "#define \fBMIN\fP(x, y)   ( (x < y) ? x : y )"
.br
.ti -1c
.RI "#define \fBABS\fP(x)   ( (x > 0) ? x : (-x) )"
.br
.ti -1c
.RI "#define \fBNULL\fP   0"
.br
.ti -1c
.RI "#define \fBUART0_TX_BUFFER_SIZE\fP   512"
.br
.ti -1c
.RI "#define \fBUART1_TX_BUFFER_SIZE\fP   512"
.br
.ti -1c
.RI "#define \fBUART2_TX_BUFFER_SIZE\fP   512"
.br
.ti -1c
.RI "#define \fBUART3_TX_BUFFER_SIZE\fP   512"
.br
.ti -1c
.RI "#define \fBUART4_TX_BUFFER_SIZE\fP   512"
.br
.ti -1c
.RI "#define \fBUART0_RX_BUFFER_SIZE\fP   512"
.br
.ti -1c
.RI "#define \fBUART1_RX_BUFFER_SIZE\fP   512"
.br
.ti -1c
.RI "#define \fBUART2_RX_BUFFER_SIZE\fP   512"
.br
.ti -1c
.RI "#define \fBUART3_RX_BUFFER_SIZE\fP   512"
.br
.ti -1c
.RI "#define \fBUART4_RX_BUFFER_SIZE\fP   512"
.br
.ti -1c
.RI "#define \fBUART_RX_FIFO_TRIGGER\fP   0"
.br
.ti -1c
.RI "#define \fBUART_LOAD_DLM\fP(div)   (((div) >> 8) & 0xFF)"
.br
.ti -1c
.RI "#define \fBUART_LOAD_DLL\fP(div)   ((div) & 0xFF)"
.br
.in -1c
.SS "Enumerations"

.in +1c
.ti -1c
.RI "enum \fBBOOLEAN\fP { \fBFALSE\fP = 0x00, \fBTRUE\fP = 0x01 }"
.br
.ti -1c
.RI "enum \fBSTATUS\fP { \fBOK\fP = 0, \fBNOT_OK\fP = 1, \fBEMPTY_QUEUE\fP = 2, \fBBUSY\fP = 3, \fBIDLE\fP = 4, \fBINCORRECT_PARAMS\fP = 5, \fBTIMEOUT\fP = 6, \fBRESERVED_6\fP = 7, \fBRESERVED_8\fP = 8, \fBNOT_INIT\fP = 9, \fBFULL_QUEUE\fP = 10, \fBEND_OF_QUEUE\fP = 11, \fBCRC_ERROR\fP = 12, \fBOUT_OF_SYNC\fP = 13, \fBINCORRECT_PASSWORD\fP = 14, \fBNO_TIMERS_AVAILABLE\fP = 15, \fBNOT_READY\fP = 16 }"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "uint8_t \fBDRV_GENERAL_GetFirstDigit\fP (uint32_t n)"
.br
.ti -1c
.RI "\fBUART_BAUDRATE_VALUE_DLD\fP \fBDRV_GENERAL_CalculateBaudrateDLD\fP (uint32_t \fBPeripheralClock\fP, uint32_t baud, uint8_t sampling)"
.br
.ti -1c
.RI "\fBUART_BAUDRATE_VALUE_DIV_MUL\fP \fBDRV_GENERAL_CalculateBaudrateDivMul\fP (uint32_t \fBPeripheralClock\fP, uint32_t baudrate, uint8_t sampling)"
.br
.in -1c
.SH "Detailed Description"
.PP 
General structures, types, functions and defines\&. 

Contains various types, functions and defines used by the rest of this driver library 
.SH "Macro Definition Documentation"
.PP 
.SS "#define ABS(x)   ( (x > 0) ? x : (-x) )"
Define macro for the absolute value of a number 
.PP
Definition at line 25 of file drv_general\&.h\&.
.SS "#define HZ   *1l"
Define for Hz 
.PP
Definition at line 21 of file drv_general\&.h\&.
.SS "#define KHZ   *1000l"
Define for KHz 
.PP
Definition at line 20 of file drv_general\&.h\&.
.SS "#define MAX(x, y)   ( (x > y) ? x : y )"
Define macro for the maximum value between two numbers 
.PP
Definition at line 23 of file drv_general\&.h\&.
.SS "#define MHZ   *1000000l"
Define for MHz 
.PP
Definition at line 19 of file drv_general\&.h\&.
.SS "#define MIN(x, y)   ( (x < y) ? x : y )"
Define macro for the minimum value between two numbers 
.PP
Definition at line 24 of file drv_general\&.h\&.
.SS "#define UART0_RX_BUFFER_SIZE   512"
Defines the receive buffer size for UART 0 
.PP
Definition at line 39 of file drv_general\&.h\&.
.SS "#define UART0_TX_BUFFER_SIZE   512"
Defines the transmit buffer size for UART 0 
.PP
Definition at line 33 of file drv_general\&.h\&.
.SS "#define UART1_RX_BUFFER_SIZE   512"
Defines the receive buffer size for UART 1 
.PP
Definition at line 40 of file drv_general\&.h\&.
.SS "#define UART1_TX_BUFFER_SIZE   512"
Defines the transmit buffer size for UART 1 
.PP
Definition at line 34 of file drv_general\&.h\&.
.SS "#define UART2_RX_BUFFER_SIZE   512"
Defines the receive buffer size for UART 2 
.PP
Definition at line 41 of file drv_general\&.h\&.
.SS "#define UART2_TX_BUFFER_SIZE   512"
Defines the transmit buffer size for UART 2 
.PP
Definition at line 35 of file drv_general\&.h\&.
.SS "#define UART3_RX_BUFFER_SIZE   512"
Defines the receive buffer size for UART 3 
.PP
Definition at line 42 of file drv_general\&.h\&.
.SS "#define UART3_TX_BUFFER_SIZE   512"
Defines the transmit buffer size for UART 3 
.PP
Definition at line 36 of file drv_general\&.h\&.
.SS "#define UART4_RX_BUFFER_SIZE   512"
Defines the receive buffer size for UART 4 
.PP
Definition at line 43 of file drv_general\&.h\&.
.SS "#define UART4_TX_BUFFER_SIZE   512"
Defines the transmit buffer size for UART 4 
.PP
Definition at line 37 of file drv_general\&.h\&.
.SS "#define UART_LOAD_DLL(div)   ((div) & 0xFF)"
Macro for loading least significant halfs of divisors 
.PP
Definition at line 89 of file drv_general\&.h\&.
.SS "#define UART_LOAD_DLM(div)   (((div) >> 8) & 0xFF)"
Macro for loading most significant halfs of divisors 
.PP
Definition at line 88 of file drv_general\&.h\&.
.SH "Enumeration Type Documentation"
.PP 
.SS "enum \fBBOOLEAN\fP"
Defines the a boolean type This structure uses DLD as the fractional divider 
.PP
\fBEnumerator: \fP
.in +1c
.TP
\fB\fIFALSE \fP\fP
False boolean value 
.TP
\fB\fITRUE \fP\fP
True boolean value 
.PP
Definition at line 54 of file drv_general\&.h\&.
.SS "enum \fBSTATUS\fP"
Defines a status return type This enum defines the possible return status types of the functions defined by this library 
.PP
\fBEnumerator: \fP
.in +1c
.TP
\fB\fIOK \fP\fP
Operation finished ok 
.TP
\fB\fINOT_OK \fP\fP
Operation finished not ok 
.TP
\fB\fIEMPTY_QUEUE \fP\fP
The queue/buffer is empty 
.TP
\fB\fIBUSY \fP\fP
The called module is busy 
.TP
\fB\fIIDLE \fP\fP
The called module is empty 
.TP
\fB\fIINCORRECT_PARAMS \fP\fP
The parameters are incorrect 
.TP
\fB\fITIMEOUT \fP\fP
Timeout occured in the called module 
.TP
\fB\fINOT_INIT \fP\fP
The module was not initialized 
.TP
\fB\fIFULL_QUEUE \fP\fP
The queue/buffer is full 
.TP
\fB\fIEND_OF_QUEUE \fP\fP
End of queue/buffer was reached 
.TP
\fB\fICRC_ERROR \fP\fP
CRC check failed 
.TP
\fB\fIOUT_OF_SYNC \fP\fP
Operation was out of sync 
.TP
\fB\fIINCORRECT_PASSWORD \fP\fP
Incorrect password 
.TP
\fB\fINO_TIMERS_AVAILABLE \fP\fP
No more software timers available 
.TP
\fB\fINOT_READY \fP\fP
The called module is not ready 
.PP
Definition at line 97 of file drv_general\&.h\&.
.SH "Function Documentation"
.PP 
.SS "\fBUART_BAUDRATE_VALUE_DIV_MUL\fP DRV_GENERAL_CalculateBaudrateDivMul (uint32_tPeripheralClock, uint32_tbaudrate, uint8_tsampling)\fC [private]\fP"
Calculate the baud rate divisors using the fractional divisor as mval/dval
.PP
This function is used to calculate de divisors for a specific baudrate and peripheral input clock 
.PP
\fBParameters:\fP
.RS 4
\fIPeripheralClock\fP The Peripheral input clock in Hz 
.br
\fIbaudrate\fP Desired baud rate 
.br
\fIsampling\fP Sampling rate (8, 16) 
.RE
.PP
\fBReturns:\fP
.RS 4
A structure containing DLM, DLL and the fractional components mval and dval 
.RE
.PP

.PP
Definition at line 104 of file drv_general\&.c\&.
.SS "\fBUART_BAUDRATE_VALUE_DLD\fP DRV_GENERAL_CalculateBaudrateDLD (uint32_tPeripheralClock, uint32_tbaud, uint8_tsampling)\fC [private]\fP"
Calculate the baud rate divisors using the DLD method
.PP
This function is used to calculate de divisors for a specific baudrate and peripheral input clock 
.PP
\fBParameters:\fP
.RS 4
\fIPeripheralClock\fP The Peripheral input clock in Hz 
.br
\fIbaud\fP Desired baud rate 
.br
\fIsampling\fP Sampling rate (8, 16) 
.RE
.PP
\fBReturns:\fP
.RS 4
A structure containing DLM, DLL and DLD 
.RE
.PP

.PP
Definition at line 61 of file drv_general\&.c\&.
.SS "uint8_t DRV_GENERAL_GetFirstDigit (uint32_tn)\fC [private]\fP"
Calculates the first digit of a number
.PP
This function is used to extract the first digit of a number 
.PP
\fBParameters:\fP
.RS 4
\fIn\fP The number to extract the 1st digit from 
.RE
.PP
\fBReturns:\fP
.RS 4
The 1st digit of the number 
.RE
.PP

.PP
Definition at line 37 of file drv_general\&.c\&.
.SH "Author"
.PP 
Generated automatically by Doxygen for Lab TD from the source code\&.


AT/AT/code/drv/drv_general.c
//*****************************************************************************
//! \file	drv_general.c
//! \author	Valentin STANGACIU, DSPLabs
//! 
//! \brief	General structures, types, functions and defines
//! 
//! Contains various types, functions and defines used by the rest of this driver library
//*****************************************************************************

//*****************************************************************************
//! \headerfile drv_general.h "drv_general.h"
//*****************************************************************************

//*****************************************************************************
//! \addtogroup DRV_GENERAL
//! @{
//! \brief	General structures, types, functions and defines
//! 
//! Contains various types, functions and defines used by the rest of this driver library
//*****************************************************************************

#include "drv_general.h"

#include <stdint.h>
#include "LPC177x_8x.h"
#include "system_LPC177x_8x.h"


//*****************************************************************************
//! Calculates the first digit of a number
//!
//! This function is used to extract the first digit of a number
//! \param n The number to extract the 1st digit from
//! \return The 1st digit of the number
//! \private
//*****************************************************************************
uint8_t DRV_GENERAL_GetFirstDigit(uint32_t n)
{
	uint32_t temp = n;
	while (temp != 0)
	{
		if ((temp / 10) == 0)
		{
			return (temp % 10);
		}
		temp = temp / 10;
	}
	return 0;
}

//*****************************************************************************
//! Calculate the baud rate divisors using the DLD method
//!
//! This function is used to calculate de divisors for a specific baudrate and peripheral input clock
//! \param PeripheralClock The Peripheral input clock in Hz
//! \param baud Desired baud rate
//! \param sampling Sampling rate (8, 16)
//! \return A structure containing DLM, DLL and DLD
//! \private
//*****************************************************************************
UART_BAUDRATE_VALUE_DLD DRV_GENERAL_CalculateBaudrateDLD(uint32_t PeripheralClock, uint32_t baud, uint8_t sampling)
{
	uint64_t temp64 = 0;
	uint32_t temp32 = 0;
	uint32_t temp;
	UART_BAUDRATE_VALUE_DLD result = {0, 0, 0};
	if ((PeripheralClock % (baud * sampling)) == 0)
	{
		// nu am nevoie de divizor fractional		
		temp32 = PeripheralClock / (baud * sampling);
		result.DLD = 0;
		result.DLM = (temp32 >> 8) & 0xFF;
		result.DLL = temp32 & 0xFF;
	}
	else
	{
		// am nevoie si de divizor fractional
		temp64 = ((uint64_t)PeripheralClock * 1000) / (baud * sampling);
		temp32 = temp64 - ((PeripheralClock / (baud * sampling)) * 1000);
		temp = PeripheralClock / (baud * sampling);
		result.DLM = (temp >> 8) & 0xFF;
		result.DLL = temp & 0xFF;
		temp = 16 * temp32;
		result.DLD = temp / 1000;
		temp = temp - ((temp / 1000) * 1000);
		if ((temp > 100) && (DRV_GENERAL_GetFirstDigit(temp) >= 5))
		{
			result.DLD++;
		}		
	}
	return result;
}

//*****************************************************************************
//! Calculate the baud rate divisors using the fractional divisor as mval/dval
//!
//! This function is used to calculate de divisors for a specific baudrate and peripheral input clock
//! \param PeripheralClock The Peripheral input clock in Hz
//! \param baudrate Desired baud rate
//! \param sampling Sampling rate (8, 16)
//! \return A structure containing DLM, DLL and the fractional components mval and dval
//! \private
//*****************************************************************************
UART_BAUDRATE_VALUE_DIV_MUL	 DRV_GENERAL_CalculateBaudrateDivMul(uint32_t PeripheralClock, uint32_t baudrate, uint8_t sampling)
{
	UART_BAUDRATE_VALUE_DIV_MUL result = {0,0,0,0};
	uint32_t uClk;
	uint32_t rate16 = 16 * baudrate;
	uint32_t dval, mval;
	uint32_t dl;
	uClk = PeripheralClock;
  // The fractional is calculated as
   // (PCLK  % (16 * Baudrate)) / (16 * Baudrate)
   // Let's make it to be the ratio
   // DivVal / MulVal
   //

   dval = uClk % rate16;
 
   // The PCLK / (16 * Baudrate) is fractional
   // => dval = pclk % rate16;
   // mval = rate16;
   // now mormalize the ratio
   // dval / mval = 1 / new_mval;
   // new_mval = mval / dval
   // new_dval = 1
   if (dval > 0)
   {
      mval = rate16 / dval;
      dval = 1;
 
      // in case mval still bigger then 4 bits
      // no adjustment require
      if (mval > 12)
      {
         dval = 0;
      }
   }
   dval &= 0xf;
   mval &= 0xf;
 
   dl = uClk / (rate16);
   
	result.DLM = (dl >> 8) & 0xFF;
   result.DLL =  dl & 0xFF;
//   result.dval = dval;
  // result.mval = mval;


	return result;
}

//*****************************************************************************
//
// Close the Doxygen group.
//! @}
//
//*****************************************************************************
